### PR TITLE
PIN-3022 [BE] Nel listing delle finalità lato erogatore devono essere escluse quelle in stato DRAFT

### PR DESCRIPTION
**Pull request by nttdata-rtorsoli**

### CODE REVIEW

File: src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/PurposesApiServiceImpl.scala [hunk 1]
```diff
@@ -85,7 +87,57 @@ final case class PurposesApiServiceImpl(
...
+    contexts: Seq[(String, String)],
+    toEntityMarshallerPurposes: ToEntityMarshaller[Purposes],
+    toEntityMarshallerProblem: ToEntityMarshaller[Problem]
+  ): Route = {
+    val result: Future[Purposes] =
+      for {
+        statesEnum     <- parseArrayParameters(states).distinct
+          .traverse(PurposeProcess.PurposeVersionState.fromValue)
+          .toFuture
+        eServicesUUIDs <- parseArrayParameters(eServicesIds).distinct.traverse(_.toFutureUUID)
```
Comment (TonioGela) [inline src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/PurposesApiServiceImpl.scala]: We should really write an util function in commons for this use case :\

File: src/main/resources/interface-specification.yml [hunk 2]
```diff
@@ -4170,6 +4243,10 @@ paths:
               $ref: '#/components/schemas/PurposeVersionState'
             default: []
           explode: false
+        - in: query
+          name: excludeDraft
+          schema:
+            type: boolean
```
Comment (galales) [inline src/main/resources/interface-specification.yml]: This should not be exposed to the frontend.
The flag is set by the endpoint when calling the process
```suggestion
```

File: src/main/resources/interface-specification.yml [hunk 3]
```diff
@@ -4224,14 +4301,81 @@ paths:
...
+          schema:
+            type: array
+            items:
+              $ref: '#/components/schemas/PurposeVersionState'
+            default: []
+          explode: false
+        - in: query
+          name: excludeDraft
+          schema:
+            type: boolean
```
Comment (galales) [inline src/main/resources/interface-specification.yml]: Same
```suggestion
```

File: src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/PurposesApiServiceImpl.scala [hunk 4]
```diff
@@ -68,6 +69,7 @@ final case class PurposesApiServiceImpl(
           consumersIds = consumersUUIDs,
           producersIds = producersUUIDs,
           states = statesEnum,
+          excludeDraft = excludeDraft,
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/PurposesApiServiceImpl.scala]:
```suggestion
          excludeDraft = false.some,
```

File: src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/PurposesApiServiceImpl.scala [hunk 5]
```diff
@@ -85,7 +87,57 @@ final case class PurposesApiServiceImpl(
...
+        eServicesUUIDs <- parseArrayParameters(eServicesIds).distinct.traverse(_.toFutureUUID)
+        consumersUUIDs <- parseArrayParameters(consumersIds).distinct.traverse(_.toFutureUUID)
+        producersUUIDs <- parseArrayParameters(producersIds).distinct.traverse(_.toFutureUUID)
+        pagedResults   <- purposeProcessService.getPurposes(
+          name = q,
+          eServicesIds = eServicesUUIDs,
+          consumersIds = consumersUUIDs,
+          producersIds = producersUUIDs,
+          states = statesEnum,
+          excludeDraft = excludeDraft,
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/PurposesApiServiceImpl.scala]:
```suggestion
          excludeDraft = true.some,
```

File: src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/PurposesApiServiceImpl.scala [hunk 6]
```diff
@@ -85,7 +87,57 @@ final case class PurposesApiServiceImpl(
     onComplete(result) {
       handleError(
         s"Error retrieving Purposes for name $q, EServices $eServicesIds, Consumers $consumersIds offset $offset, limit $limit"
-      ) orElse { case Success(r) => getPurposes200(r) }
+      ) orElse { case Success(r) => getConsumerPurposes200(r) }
+    }
+  }
+
+  override def getProducerPurposes(
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/PurposesApiServiceImpl.scala:67]: the two endpoints are identical, let's write a single parametric function called by both and remove code duplication

**Review by galales — CHANGES_REQUESTED** [review]

**Review by galales — APPROVED** [review]

**Review by beetlecrunch — APPROVED** [review]

File: src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/PurposesApiServiceImpl.scala (hunk 1 above)
Comment (nttdata-rtorsoli) [reply src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/PurposesApiServiceImpl.scala]: We can do that in other specific story

**Pull request merged**